otherlibs/systhreads/st_stubs.c

```diff
@@ -103,20 +102,21 @@ struct caml_thread_table {
 /* thread_table instance, up to Max_domains */
 static struct caml_thread_table thread_table[Max_domains];
 
-/* the "head" of the circular list of thread descriptors for this domain */
-#define All_threads thread_table[Caml_state->id].all_threads
+#define Dom_id (this_thread->domain_id)
 
-/* The descriptor for the currently executing thread for this domain */
-#define Current_thread thread_table[Caml_state->id].current_thread
+/* The descriptor for the currently executing thread for this domain;
+   also the head of a circular list of thread descriptors for this
+   domain. */
+#define Current_thread thread_table[Dom_id].current_thread
```

Comment:
All of these macros now use `this_thread` rather than `Caml_state`. I am trying to convince myself that this change does not alter the existing behaviour. Can you explain how?

Also, `Caml_state` is shared by both the main thread and the backup thread. Hence, `Caml_state->id` has the same value if called from main thread and the backup thread. I assume that `this_thread` is not initialized on the backup thread. Are we sure that any function that uses `Dom_id` will never be called on the backup thread?

Reply:
- this_thread->id is initialized to the domain state's id. The few functions that initialize/reset this_thread had to move around some calls so the macros are called while this_thread is defined (easy to check that it is equivalent).
- Indeed, I checked that no other function is unexpectedly called from the backup thread (for the vast majority it would not make sense in the first place).